File: python/comUn.py
# kb 2-7-2025
# simulator to experiment with Lychgate RPG combat scenarios
# combat unit class
import random

class comUn:
    def __init__(self, name, idr, hp, hitChance, dmg, ac, dodge, ini):
        self.name = name
        self.idr = int(idr)
        self.hp = int(hp)
        self.maxHp = self.hp
        self.hitChance = float(hitChance)
        self.dmg = int(dmg)
        self.ac = int(ac)
        self.dodge = float(dodge)
        self.ini = int(ini)
        self.alive=1
        self.ready=1


#hp functions
    def checkDeath(self):
        if(self.hp<=0):
            print(self.name, "has fallen to their wounds.")
            self.alive=0
            self.ready=0
            return 1
        else:
            return 0
    def setHp(self, inHp):
        self.hp = inHp
        return self.alive
        
    def subHp(self, inHp):
        self.hp -= inHp
        reportHp = self.hp
        self.checkDeath()
        return reportHp
        
    def addHp(self, inHp):
        self.hp += inHp
        reportHp = self.hp
        return reportHp
    
    # combat logic
    def tryDodge(self, inShot):
        #react formula: super basic
        react=self.dodge
        
        if (inShot>react):
            # the dodge fails
            return 0
        else:
            # the dodge succeeds
            return 1
        
        
    def attack(self, inTarget):
        #first we need to hit
        #grab our accuracy and damage
        acc = self.hitChance
        dmg = self.dmg
        
        # TODO : imeplement new hit/dodge algorithm
        
        
        #roll a D100 to swing at the target
        roll = random.randrange(1,100)
        # add the roll value to our base accuracy
        netAcc = acc + roll
        thresh = inTarget.dodge
        #
        if (netAcc > thresh):
            netDmg = dmg - inTarget.ac
            inTarget.subHp(netDmg)
            print(self.name, "deals", netDmg, "to", inTarget.name)
            
            #return 1 true for successful hit
            return netDmg
       
        else:
            #report a miss
            print(self.name, "swings but", inTarget.name, " is too quick!")
            #return 0 false for miss
            return 0
    
    
    def exhaust(self):
        self.ready=0
        return 1
        
class Hero(comUn):
    def __init__(self, name, idr, hp, hitChance, dmg, ac, dodge, ini): 
        super().__init__(name, idr, hp, hitChance, dmg, ac, dodge, ini)
        #distinguish player
        self.side="player"
        
        
class Enemy(comUn):
    def __init__(self, name, idr, hp, hitChance, dmg, ac, dodge, ini): 
        super().__init__(name, idr, hp, hitChance, dmg, ac, dodge, ini)
        #distinguish enemy
        self.side="enemy"

File: dotnet/ComUn.cs

using MathNet.Numerics;
using MathNet.Numerics.Distributions;

namespace Lychgate
{
    //faction enumeration
    public enum Faction
    {

        Heroes,
        Enemies

    }


    //base combat unit class
    public class ComUn
    {
        // instance variables
        public string Name;
        public int Idr;
        public int Hp;
        public int MaxHp;
        public double Accuracy;
        public int Dmg;
        public int Ac;
        public double Dodge;
        public int Ini;
        public bool Alive;
        public bool Ready;
        public Faction Alignment;
        public string strAlignTag = ""; //faction represented as string. needed to make the JSON parsing simpler.

        public ComUn(string name, int idr, int hp, double hitChance, int dmg, int ac, double dodge, int ini)
        {
            Name = name;
            Idr = idr;
            Hp = hp;
            MaxHp = hp;
            Accuracy = hitChance;
            Dmg = dmg;
            Ac = ac;
            Dodge = dodge;
            Ini = ini;
            Alive = true;
            Ready = true;
        }

        [Newtonsoft.Json.JsonConstructor]
        public ComUn(string name, int idr, int hp, double hitChance, int dmg, int ac, double dodge, int ini, string align)
        {
            //todo set alignment
            switch (align) {
                case "Hero":
                    this.Alignment = Faction.Heroes;
                    break;
                case "Mob":
                    this.Alignment = Faction.Enemies;
                    break;
                default:
                    throw new Exception("Invalid alignment value: " + align);
            
                }
            Name = name;
            Idr = idr;
            Hp = hp;
            MaxHp = hp;
            Accuracy = hitChance;
            Dmg = dmg;
            Ac = ac;
            Dodge = dodge;
            Ini = ini;
            Alive = true;
            Ready = true;
        }

        public static void CreateUnit()
        {   //make sure this level is working.
            Console.WriteLine("CreateUnit is working.");
        }
        
        public bool CheckDeath()
        {   //is the ceeature alive? TODO: Does this need to pass the creature ID?
            if (Hp <= 0)
            {
                
                Alive = false;
                Ready = false;
                return true;
            }
            else
            {
                return false;
            }
            //Console.WriteLine("CheckDeath is working.");
        }
        public bool SetHP(int inHP)
        {
            Hp = inHP;
            return Alive;
        }
        public int AddHP(int addHp)
        {
            Hp = Hp + addHp;
            return Hp;
        }
        public int SubHP(int minusHp)
        {
            Hp = Hp - minusHp;
            return Hp;
        }

        public void ReadyUp()
        {
            if (Alive)
            {
                this.Ready = true;
            }
        }

        //todo: implement Attack() method

        private bool TryHit(double acc, double targetDodge)
        {
            Random d100 = new();
            double roll = d100.NextDouble();

            roll += acc;

            if (roll > targetDodge)
            {
                return true;

            }
            else
            {
                return false;

            }

        }

        public int Attack(ComUn inTarget)
        {
            //roll D100
            bool hit = TryHit(this.Accuracy, inTarget.Dodge);
            if (hit)
            {   
                // Roll Damage 

                int netDmg = RollDmg() - inTarget.Ac;
                if (netDmg < 0)
                {
                    // don't return less than 0 if it doesnt get thru armor
                    return 0;
                }
                else
                {
                    // successful hit, roll damage
                    return netDmg;
                }
            }
            else
            {
                //unique error code if a total miss; net dmg can be zero.
                return -1;
            }

        }
        private int RollDmg()
        {
            double rolled_dmg = (int) Normal.Sample(this.Dmg, 2.0);

            if (rolled_dmg <= 0)
            {
                rolled_dmg = 0;
            }

            Console.WriteLine($"Damage Roll: {rolled_dmg}");
            return (int) rolled_dmg;

        }

        public void Exhaust()
        {
            //unit has taken its turn for the round.
            this.Ready = false;
        }
    }


    //todo: extend class to Hero and Enemy subclasses

    public class Hero : ComUn
    {

        public Hero(string name, int idr, int hp, double hitChance, int dmg, int ac, double dodge, int ini) : base(name, idr, hp, hitChance, dmg, ac, dodge, ini)
        {
            Alignment = Faction.Heroes;
        }
    }
    public class Mob : ComUn
    {

        public Mob(string name, int idr, int hp, double hitChance, int dmg, int ac, double dodge, int ini) : base(name, idr, hp, hitChance, dmg, ac, dodge, ini)
        {
            Alignment = Faction.Enemies;
        }
    }

}